### PR TITLE
[inductor] Inline ComputedBuffer computation when there are no reads

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102277
* __->__ #102000

When inductor compiles the following example,
```python
def flip(x):
    idx = torch.arange(x.shape[0] - 1, -1, -1, device=x.device)
    return x[idx], idx
```

The return of `idx` forces it to be realized into a `ComputedBuffer`
and the downstream index call inserts a corresponding load and
indirect_indexing:
```python
    tmp0 = tl.load(in_ptr0 + (x1), None)
    tmp1 = triton_helpers.promote_to_tensor(tmp0)
    tl.device_assert((0 <= tmp1) & (tmp1 < 128), "index out of bounds: 0 <= tmp1 < 128")
    tmp2 = tl.load(in_ptr1 + (x0 + (128*tmp0)), None)
```

However, if we can inline the index expression from the buffer's
computation we instead get direct indexing (and half the loads):
```python
    tmp0 = tl.load(in_ptr0 + (127 + ((-1)*x0)), None)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -1989,12 +1997,16 @@ def unwrap_views(target):
                 return unwrap_views(target.data)
             return target
 
-        return unwrap_views(self.target).layout
+        result = unwrap_views(self.target)
+        assert isinstance(result, Buffer), "MutationLayout must refer to a buffer"
+        return result
```
Comment (peterbell10) [inline torch/_inductor/ir.py:2002]: I was seeing accuracy failures and narrowed it down to certain fallback kernels like scatter fallback not calling `realize_users_of`, so this strictly enforces that creating a `MutationLayout` will mark the buffer as mutated.

**Review by lezcano — APPROVED** [review]
Great stuff. I also like a lot the split between reindexing and making the actual index to allow you to work with the indices on their own.

**Comment by kit1980** [comment]
See comments to the next PR in the stack https://github.com/pytorch/pytorch/pull/102277 for more details on the revert.

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -5185,7 +5238,7 @@ def run(x):
             self.assertEqual(bw_code.count("tl.rand"), 0)
             expected_kernel = 4
         else:
-            expected_kernel = 5
+            expected_kernel = 6
```
[the remark this answers is not in the document]
Comment (lezcano) [reply test/inductor/test_torchinductor.py:5241]: so this is an example of the usual candidate to be fused with `aggressive_fusion = True`, but that we're not fusing ATM. Fair enough.

**Comment by ezyang** [comment]
I'm guessing this is responsible for a 2% HF overall improvement. Example

image

Other candidate PRs:

```
(base) ezyang-mbp:pytorch-tmp ezyang$ git log --oneline af1d437654aa7fd5aa4a5d30ae2f2e9af34c7765..999bae0f54108ffc5b7cf2524a02a83901554b16 torch/_dynamo torch/_inductor .ci
2cc6ae1926d (origin/gh/willfengg/1/base) squash xblock for persistent inner reduction (#102444)
fc0fed36d93 [inductor] fix issue with ops.lookup_seed (#102485)
c6d9a0b9dd4 [inductor] Handle floordiv and remainder in IndexPropagation (#102277)
e4e151d6693 [inductor] Inline ComputedBuffer computation when there are no reads (#102000)
b1bc8aecf5b [inductor] erfinv: CPU/CUDA lowering (#101863)
```

**Comment by ezyang** [comment]
Hmm, actually, also decent chance this is due to https://github.com/pytorch/pytorch/pull/102444

**Pull request closed without merging**